### PR TITLE
[RFC] api: implement nvim_buf_get_offset_for_line

### PR DESCRIPTION
**Pull request by bfredl**
Like line2byte, but works for any buffer, and uses zero-based indexing (API conventions).

Needed for efficient tree-sitter bindings, ref https://github.com/tree-sitter/tree-sitter/issues/210

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -491,6 +491,37 @@ void nvim_buf_set_lines(uint64_t channel_id,
...
+Integer nvim_buf_get_offset_for_line(Buffer buffer, Integer index, Error *err)
+  FUNC_API_SINCE(5)
+{
+  buf_T *buf = find_buffer_by_handle(buffer, err);
+  if (!buf) {
+    return 0;
+  }
+
+  if (index < 0 || index > buf->b_ml.ml_line_count) {
+    api_set_error(err, kErrorTypeValidation, "Index out of bounds");
```
Comment (justinmk) [inline src/nvim/api/buffer.c]: Should we also return a sentinel for unloaded buffer? https://github.com/neovim/neovim/blob/ff5ab82c55be76338f8bdfbff8047796c7b33e3c/src/nvim/api/buffer.c#L247

**Review by justinmk — APPROVED** [review]

File: src/nvim/api/buffer.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/api/buffer.c]: Done so.

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -491,6 +491,43 @@ void nvim_buf_set_lines(uint64_t channel_id,
...
+/// Return the byte offset for a line.
+//
+/// This includes the end-of-line character, depending on the 'fileformat'
+/// option for the current buffer.  The first line returns 0.
+/// UTF-8 encoding is used, 'fileencoding' is ignored.
+/// Sending in the index just below the last line gives
+/// the total byte count for the entire file. A final newline
+/// will be included if it would be written, see |eol|.
+///
+/// Unlike |line2byte|, throws error for out-of-bounds indexing.
```
Comment (justinmk) [inline src/nvim/api/buffer.c]: Technically the tag is `|line2byte()|`. `make html` will complain about this, I fixed several in add2a6239810a936f682d340525ac98e5e1fb15b.

For the same reason `|eol|` might need to be `|'eol'|` or just `'eol'`.

**Comment by justinmk** [comment]
seems to fail on appveyor

```
  [  FAILED  ] C:/projects/neovim/test/functional\api\buffer_spec.lua @ 306: api/buf get_offset_for_line works
  C:/projects/neovim/test/functional\api\buffer_spec.lua:310: Expected objects to be the same.
  Passed in:
  (number) 6
  Expected:
  (number) 5
```

**Comment by bfredl** [comment]
Hmm, maybe `nvim_buf_get_offset_for_line` should just ignore `fileformat`.

**Comment by justinmk** [comment]
Any objection to name `nvim_buf_get_offset` ?

**Pull request merged**